servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/CacheConnectionFactory.java

```diff
@@ -0,0 +1,177 @@
+    }
+
+    @SuppressWarnings("deprecation")
+    @Deprecated
+    @Override
+    public Single<C> newConnection(final ResolvedAddress resolvedAddress,
+                                   @Nullable final TransportObserver observer) {
+        return newConnection(resolvedAddress, null, observer);
+    }
+
+    @Override
+    public Single<C> newConnection(final ResolvedAddress resolvedAddress, @Nullable final ContextMap context,
+                                   @Nullable final TransportObserver observer) {
+        return Single.defer(() -> {
+            final int maxConcurrency = maxConcurrencyFunc.applyAsInt(resolvedAddress);
+            if (maxConcurrency <= 1) {
+                // If the concurrency is <=1 there is no use in caching the connection because we won't re-use it.
+                return delegate().newConnection(resolvedAddress, context, observer);
+            }
+
+            Single<C> result;
+            synchronized (map) {
+                final Item<C> item1 = map.get(resolvedAddress);
+                if (item1 == null || (result = item1.addSubscriber(maxConcurrency)) == null) {
+                    final Item<C> item2 = new Item<>();
+                    map.put(resolvedAddress, item2);
+                    result = item2.single = delegate().newConnection(resolvedAddress, context, observer)
+                            // Remove from map in a cancel above cache operator. Cache will only cancel upstream
+                            // after all subscribers have cancelled.
+                            .<C>liftSync(subscriber -> new Subscriber<C>() {
+                                @Override
+                                public void onSubscribe(final Cancellable cancellable) {
+                                    subscriber.onSubscribe(() -> {
+                                        try {
+                                            assert Thread.holdsLock(map);
+                                            map.remove(resolvedAddress, item2);
+                                        } finally {
+                                            cancellable.cancel();
+                                        }
+                                    });
+                                }
+
+                                @Override
+                                public void onSuccess(@Nullable final C result1) {
+                                    try {
+                                        if (result1 == null) {
+                                            lockRemoveFromMap();
+                                        } else {
+                                            result1.onClose().whenFinally(this::lockRemoveFromMap).subscribe();
+                                        }
+                                    } catch (Throwable cause) {
+                                        if (result1 != null) {
+                                            result1.closeAsync().subscribe();
```

Comment:
Consider logging (at least at debug level) the fact we close a connection with the cause